kt/godot-library/src/main/kotlin/godot/core/Projection.kt

```diff
@@ -0,0 +1,1128 @@
+        columns[0].z = 0.0
+        columns[0].w = 0.0
+        columns[1].x = 0.0
+        columns[1].y = p_rect.size.y
+        columns[1].z = 0.0
+        columns[1].w = 0.0
+        columns[2].x = 0.0
+        columns[2].y = 0.0
+        columns[2].z = 1.0
+        columns[2].w = 0.0
+        columns[3].x = p_rect.position.x
+        columns[3].y = p_rect.position.y
+        columns[3].z = 0.0
+        columns[3].w = 1.0
+    }
+
+    fun setPerspective(pFovyDegrees: RealT, aspect: RealT, zNear: RealT, zFar: RealT, flipFov: Boolean) {
+        val fovyDegrees = if (flipFov) getFovy(pFovyDegrees, 1.0 / aspect) else pFovyDegrees
+
+        val radians = Math.toRadians(fovyDegrees / 2.0)
+        val sine: RealT = sin(radians)
+        val deltaZ = zFar - zNear
+
+        if (deltaZ == 0.0 || sine == 0.0 || aspect == 0.0) {
+            return
+        }
+
+        val cotangent = cos(radians) / sine
+
+        setIdentity()
+
+        columns[0][0] = cotangent / aspect
+        columns[1][1] = cotangent
+        columns[2][2] = -(zFar + zNear) / deltaZ
+        columns[2][3] = -1.0
+        columns[3][2] = -2 * zNear * zFar / deltaZ
+        columns[3][3] = 0.0
+    }
+
+    fun setPerspective(
+        pFovyDegrees: RealT,
+        aspect: RealT,
+        zNear: RealT,
+        zFar: RealT,
+        flipFov: Boolean,
+        eye: Int,
+        intraocularDist: RealT,
+        convergenceDist: RealT
+    ) {
+        val fovyDegrees = if (flipFov) getFovy(pFovyDegrees, 1.0 / aspect) else pFovyDegrees
+
+        val ymax = zNear * tan(Math.toRadians(fovyDegrees / 2.0))
+        val xmax = ymax * aspect
+        val frustumshift = (intraocularDist / 2.0) * zNear / convergenceDist
+
+        val left: RealT
+        val right: RealT
+        val modelTranslation: RealT
+        when (eye) {
+            1 -> {
+                left = -xmax + frustumshift
+                right = xmax + frustumshift
+                modelTranslation = intraocularDist / 2.0
+            }
+
+            2 -> {
+                left = -xmax - frustumshift
+                right = xmax - frustumshift
+                modelTranslation = -intraocularDist / 2.0
+            }
+
+            else -> {
+                left = -xmax
+                right = xmax
+                modelTranslation = 0.0
+            }
+        }
+
+        setFrustrum(left, right, -ymax, ymax, zNear, zFar)
+
+        val cm = Projection()
+        cm.columns[3].x = modelTranslation
+        set(this * cm)
+    }
+
+    fun setForHmd(
+        pEye: Int,
+        pAspect: RealT,
+        pIntraocularDist: RealT,
+        pDisplayWidth: RealT,
+        pDisplayToLens: RealT,
+        pOversample: RealT,
+        pZNear: RealT,
+        pZFar: RealT
+    ) {
+        // we first calculate our base frustum on our values without taking our lens magnification into account.
+        // we first calculate our base frustum on our values without taking our lens magnification into account.
```

Comment:
Duplicate

Reply:
Yes, there are some duplicates, but I feel like if we remove them we will have troubles to maintain the code. I kept it the closer possible to cpp code.